src/test/java/api/loans/ScheduledDueDateNoticesProcessingTests.java

```diff
@@ -0,0 +1,385 @@
+          BEFORE_TIMING, beforeTemplateId,
+          beforeRecurringPeriod, true)));
+    }
+    if (uponAtNoticeNextRunTime != null) {
+      assertThat(scheduledNotices, hasItems(
+        hasScheduledLoanNotice(
+          loan.getId(), uponAtNoticeNextRunTime,
+          UPON_AT_TIMING, uponAtTemplateId,
+          null, true)));
+    }
+    if (afterNoticeNextRunTime != null) {
+      assertThat(scheduledNotices, hasItems(
+        hasScheduledLoanNotice(
+          loan.getId(), afterNoticeNextRunTime,
+          AFTER_TIMING, afterTemplateId,
+          afterRecurringPeriod, true)));
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  private void checkSentNotices(UUID... expectedTemplateIds)
+    throws MalformedURLException,
+    InterruptedException,
+    TimeoutException,
+    ExecutionException {
+
+    Map<String, Matcher<String>> noticeContextMatchers = new HashMap<>();
+    noticeContextMatchers.putAll(NoticeMatchers.getUserContextMatchers(borrower));
+    noticeContextMatchers.putAll(NoticeMatchers.getItemContextMatchers(item));
+    noticeContextMatchers.putAll(NoticeMatchers.getLoanContextMatchers(loan, 0));
+    noticeContextMatchers.putAll(NoticeMatchers.getLoanPolicyContextMatchers(
+      loanPoliciesFixture.canCirculateRolling(), 0));
+
+    Matcher[] matchers = Stream.of(expectedTemplateIds)
+      .map(templateId -> hasEmailNoticeProperties(borrower.getId(), templateId, noticeContextMatchers))
+      .toArray(Matcher[]::new);
+
+    List<JsonObject> sentNotices = patronNoticesClient.getAll();
+    assertThat(sentNotices, hasSize(expectedTemplateIds.length));
+    assertThat(sentNotices, hasItems(matchers));
+  }
+
+  @Test
+  public void beforeNoticeShouldBeSentAndItsNextRunTimeShouldBeUpdated()
+    throws MalformedURLException,
+    InterruptedException,
+    TimeoutException,
+    ExecutionException {
+
+    DateTime beforeDueDateTime = dueDate.minus(beforePeriod.timePeriod()).plusSeconds(1);
+    scheduledNoticeProcessingClient.runNoticesProcessing(beforeDueDateTime);
+
+    checkSentNotices(beforeTemplateId);
+
+    DateTime expectedNewRunTimeForBeforeNotice = dueDate
+      .minus(beforePeriod.timePeriod())
+      .plus(beforeRecurringPeriod.timePeriod());
+
+    checkScheduledNotices(
+      expectedNewRunTimeForBeforeNotice,
+      dueDate,
+      dueDate.plus(afterPeriod.timePeriod()));
+  }
+
+  @Test
+  public void beforeNoticeShouldBeSendAndDeletedWhenItsNextRunTimeIsAfterDueDate()
+    throws MalformedURLException,
+    InterruptedException,
+    TimeoutException,
+    ExecutionException {
+
+    DateTime justBeforeDueDateTime = dueDate.minusSeconds(1);
+    scheduledNoticeProcessingClient.runNoticesProcessing(justBeforeDueDateTime);
+
+    checkSentNotices(beforeTemplateId);
+
+    checkScheduledNotices(
+      null,
+      dueDate,
+      dueDate.plus(afterPeriod.timePeriod()));
+  }
+
+  @Test
+  public void beforeAndUponAtNoticesShouldBeSentWhenProcessingJustAfterDueDate()
```

Comment:
If processing hasn't occurred till after the due date has passed, why do we still send the before notice?

As I understand it, these are typically reminder prompts, and no longer of any use after the due date has passed.

Is that understanding sensible? Has this behaviour been checked with Darcy?

Reply:
I'm mostly curious, this question won't block the merging of this pull request

Reply:
I agree, sending before notice has no sense when due date is passed.
I've made an appropriate change.